flink-ml-lib/src/main/java/org/apache/flink/ml/feature/vectorassembler/VectorAssembler.java

```diff
@@ -0,0 +1,174 @@
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Vector assembler is a transformer that combines a given list of columns into a single vector
+ * column. It will combine raw features and features generated by different feature transformers
+ * into a single feature vector. The input features of this transformer must be a vector feature or
+ * a numerical feature.
+ */
+public class VectorAssembler
+        implements Transformer<VectorAssembler>, VectorAssemblerParams<VectorAssembler> {
+    private final Map<Param<?>, Object> paramMap = new HashMap<>();
+    private static final double RATIO = 1.5;
+
+    public VectorAssembler() {
+        ParamUtils.initializeMapWithDefaultValues(paramMap, this);
+    }
+
+    @Override
+    public Table[] transform(Table... inputs) {
+        Preconditions.checkArgument(inputs.length == 1);
+
+        StreamTableEnvironment tEnv =
+                (StreamTableEnvironment) ((TableImpl) inputs[0]).getTableEnvironment();
+        RowTypeInfo inputTypeInfo = TableUtils.getRowTypeInfo(inputs[0].getResolvedSchema());
+        RowTypeInfo outputTypeInfo =
+                new RowTypeInfo(
+                        ArrayUtils.addAll(
+                                inputTypeInfo.getFieldTypes(), TypeInformation.of(Vector.class)),
+                        ArrayUtils.addAll(inputTypeInfo.getFieldNames(), getOutputCol()));
+        DataStream<Row> output =
+                tEnv.toDataStream(inputs[0])
+                        .map(
+                                (MapFunction<Row, Row>)
+                                        row -> {
+                                            Object[] objects = new Object[getInputCols().length];
+                                            for (int i = 0; i < objects.length; ++i) {
+                                                objects[i] = row.getField(getInputCols()[i]);
+                                            }
+                                            return Row.join(
+                                                    row,
+                                                    Row.of(assemble(objects, getHandleInvalid())));
+                                        },
+                                outputTypeInfo);
+        Table outputTable = tEnv.fromDataStream(output);
+        return new Table[] {outputTable};
+    }
+
+    @Override
+    public void save(String path) throws IOException {
+        ReadWriteUtils.saveMetadata(this, path);
+    }
+
+    public static VectorAssembler load(StreamExecutionEnvironment env, String path)
+            throws IOException {
+        return ReadWriteUtils.loadStageParam(path);
+    }
+
+    @Override
+    public Map<Param<?>, Object> getParamMap() {
+        return paramMap;
+    }
+
+    private static Vector assemble(Object[] objects, String handleInvalid) {
+        int offset = 0;
+        Map<Integer, Double> map = new HashMap<>(objects.length);
```

Comment:
How about using a `TreeMap` or `LinkedHashMap` here? When converting the map to `SparseVector`, this could help keeping the indexes from `map.entrySet()` in order.

Reply:
OK, It's better.